docs/proposals/network-policy.md

```diff
@@ -0,0 +1,259 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# NetworkPolicy
+
+## Abstract
+
+A proposal for implementing a new resource - NetworkPolicy - which
+will enable definition of ingress policies for selections of pods.
+
+The design for this proposal has been created by, and discussed
+extensively within the Kubernetes networking SIG.  It has been implemented
+and tested using Kubernetes API extensions by various networking solutions already.
+
+In this design, users can create various NetworkPolicy objects which select groups of pods and
+define how those pods should be allowed to communicate with each other.  The
+implementation of that policy at the network layer is left up to the
+chosen networking solution.
+
+> Note that this proposal does not yet include egress / cidr-based policy, which is still actively undergoing discussion in the SIG.  In general,
+egress is expected to augment this proposal rather than modify it, with one exception (noted below).
+
+## Implementation
+
+The implmentation in Kubernetes consists of:
+- A v1beta NetworkPolicy API object
+- A field on the `Namespace` object to control isolation, to be developed as an annotation for now, called `net.beta.kubernetes.io/network-isolation`.
+
+### Namespace changes
+
+> Note that this section is subject to some change with the addition of egress policy as mentioned above.
+
+The `Namespace` object can be augmented with the annotation `net.beta.kubernetes.io/network-isolation` which will take one of the following values:
```

Comment:
With regards to the naming and behavior of this annotation (which caused some discussion on the last call).

I think it might be beneficial to think of this field as a "default policy" which gets applied to all pods in a namespace, rather than a feature flag for enabling or disabling policy on a given namespace.

Valid values could then be:
```
// Allow all traffic.
net.beta.kubernetes.io/default-policy = allow-all

// Deny ingress traffic, allow egress traffic.
net.beta.kubernetes.io/default-policy = deny-ingress
```

I think this matches more closely the behavior that we're defining here (if traffic doesn't match a NetworkPolicy, the default policy is applied).  This also makes this field more extensible to things like egress policy when it comes along. e.g egress control would simply be a matter of extending the set of values for this field:
```
// Deny all egress, allow all ingress.
net.beta.kubernetes.io/default-policy = deny-egress

// Deny all ingress and deny all egress.
net.beta.kubernetes.io/default-policy = deny-all
```